[SPARK-9202] capping maximum number of executor&driver information kept in Worker

CodingCat opened the pull request:
https://issues.apache.org/jira/browse/SPARK-9202

CodingCat committed:
trimFinishedExecutorsAndDrivers

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

CodingCat committed:
application is fine...

CodingCat committed:
applications are fine

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38649 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38649/consoleFull) for   PR 7714 at commit [`55c4de4`](https://github.com/apache/spark/commit/55c4de4180fee25dc6a9d9d5635b62403c5319ed).

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #38649 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38649/console) for   PR 7714 at commit [`55c4de4`](https://github.com/apache/spark/commit/55c4de4180fee25dc6a9d9d5635b62403c5319ed).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
In principle, doesn't the Master also have similar problems with retained applications?

JoshRosen commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -614,6 +621,22 @@ private[worker] class Worker(
+
+  private def trimFinishedExecutorsIfNecessary(): Unit = {
+    if (finishedExecutors.size > retainedExecutors) {
+      finishedExecutors.take(math.max(finishedExecutors.size / 10, 1)).foreach{
```
Minor style nit: space after `foreach`

rmarsch replied:
I was about to work on this same issue until finding this PR already posted. I had one observation that I wanted to bring up for discussion. `scala.collection.concurrent.HashMap` does not preserve insertion ordering when you perform operations on it for iterator or traversal methods. So `take`, for example may be able to remove recent additions where the user might prefer to just lose the oldest executors from the list. [LinkedHashMap](http://www.scala-lang.org/api/2.11.5/index.html#scala.collection.mutable.LinkedHashMap) does provide the guarantee on insertion ordering being preserved for its operations. It comes at the cost of more memory overhead to provide this guarantee, but it may be worth it.

CodingCat replied:
Good point, I agree with you

JoshRosen commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -614,6 +621,22 @@ private[worker] class Worker(
+  private def trimFinishedExecutorsIfNecessary(): Unit = {
+    if (finishedExecutors.size > retainedExecutors) {
+      finishedExecutors.take(math.max(finishedExecutors.size / 10, 1)).foreach{
+        case (executorId, _) => finishedExecutors.remove(executorId)}
```
Minor style nit: the closing `}` should be on its own line.

JoshRosen commented:
The basic approach looks okay to me, so this is on the right track.  Thanks for choosing to work on this!

CodingCat committed:
test cases and docs

AmplabJenkins commented:
Merged build triggered.

CodingCat committed:
add license info & stylistic fix

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38750 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38750/consoleFull) for   PR 7714 at commit [`8d0729e`](https://github.com/apache/spark/commit/8d0729ebc9a71d46e9792395945c19f8465995b2).

CodingCat commented:
@JoshRosen , I just updated the patch addressing your comments and added test cases and docs...for Master, I think we have something capping the memory footprint there, e.g. https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L779 and https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L930

SparkQA commented:
[Test build #38750 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38750/console) for   PR 7714 at commit [`8d0729e`](https://github.com/apache/spark/commit/8d0729ebc9a71d46e9792395945c19f8465995b2).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

CodingCat committed:
styistic fix and respect insert ordering

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38758 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38758/consoleFull) for   PR 7714 at commit [`fb3ebb7`](https://github.com/apache/spark/commit/fb3ebb7b92bab7ebb59e5742aa97a8ed0f25d1cf).

SparkQA commented:
[Test build #38758 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38758/console) for   PR 7714 at commit [`fb3ebb7`](https://github.com/apache/spark/commit/fb3ebb7b92bab7ebb59e5742aa97a8ed0f25d1cf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

CodingCat committed:
fix JsonProtocolSuite

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38769 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38769/consoleFull) for   PR 7714 at commit [`8142028`](https://github.com/apache/spark/commit/81420280bf72acfc229cc0c37125ad5dcd1c8b69).

SparkQA commented:
[Test build #38769 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38769/console) for   PR 7714 at commit [`8142028`](https://github.com/apache/spark/commit/81420280bf72acfc229cc0c37125ad5dcd1c8b69).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

CodingCat committed:
stylistic fix

SparkQA commented:
[Test build #38788 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38788/consoleFull) for   PR 7714 at commit [`000578d`](https://github.com/apache/spark/commit/000578df680e960d53ea43d1c59c54e7a3c1d68e).

SparkQA commented:
[Test build #38788 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38788/console) for   PR 7714 at commit [`000578d`](https://github.com/apache/spark/commit/000578df680e960d53ea43d1c59c54e7a3c1d68e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

JoshRosen commented:
I think that this is an important feature to get in for 1.5.0 but my review bandwidth is a little limited right now during the release crunch mode.  Therefore, I'm going to try to ping a bunch of other folks to see if any of them have spare cycles to help review. @zsxwing @srowen @sarutak, do any of you have time to take an initial pass on this PR? I'll take a look tomorrow but just wanted to see if I could get some additional eyes on this while I'm asleep :)

srowen commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -614,6 +585,74 @@ private[worker] class Worker(
+    }
+  }
+
+  private[worker] def handleDriverStateChanged(driverStateChanged: DriverStateChanged): Unit = {
```
Just to clarify, this and `handleExecutorStateChanged` are just the result of moving code (and adding the call to trim), and there aren't other changes?

CodingCat replied:
yes, just encapsulate it for easy test

JoshRosen replied:
This is a nice change

srowen commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -614,6 +585,74 @@ private[worker] class Worker(
+
+  private def trimFinishedExecutorsIfNecessary(): Unit = {
+    if (finishedExecutors.size > retainedExecutors) {
+      finishedExecutors.take(math.max(finishedExecutors.size / 10, 1)).foreach {
```
The logic looks correct. I suppose you can take from the `keys()` rather than take pairs and match on them, but this is fine. 

Is this going to cause a `ConcurrentModificationException` though? you generally can't do this in Java and this is a JDK class. Does it work because take definitely makes a copy? Tests will tell us.

Is there in general a thread-safety problem here? I don't think the code tries to protect this data structure to begin with, so maybe it's not actually a problem.

CodingCat replied:
>The logic looks correct. I suppose you can take from the keys() rather than take pairs and match on them, but this is fine.

yes

> Is this going to cause a ConcurrentModificationException though? you generally can't do this in Java and this is a JDK class. Does it work because take definitely makes a copy? Tests will tell us.

I checked this when writing the patch, Scala implements `take(n)` as 

```scala

override /*TraversableLike*/ def take(n: Int): Repr = {
    val b = newBuilder //returns a empty map

    if (n <= 0) b.result()
    else {
      b.sizeHintBounded(n, this)
      var i = 0
      val it = iterator
      while (i < n && it.hasNext) {
        b += it.next
        i += 1
      }
      b.result()
    }
  }

```

so, I thought it's OK 

> Is there in general a thread-safety problem here? I don't think the code tries to protect this data structure to begin with, so maybe it's not actually a problem.

Hmm... I checked that when I wrote these lines, it seems that all operations are done by the actor (now RpcEndPoint), as long as we ensure the thread-safety of a single RpcEndPoint, it should be fine

JoshRosen replied:
Aren't these data structures also read by the web ui?  Do those reads go through the actor?

JoshRosen replied:
If this really is safe then we should leave a line comment in the code to record our rationale

CodingCat replied:
yes, the current design ensures that the actor is the channel getting WorkerStates

https://github.com/apache/spark/blob/069a4c414db4612d7bdb6f5615c1ba36998e5a49/core/src/main/scala/org/apache/spark/deploy/worker/ui/WorkerPage.scala#L36

https://github.com/apache/spark/blob/069a4c414db4612d7bdb6f5615c1ba36998e5a49/core/src/main/scala/org/apache/spark/deploy/worker/ui/WorkerPage.scala#L41

I'm adding the comments

JoshRosen replied:
Great, thanks for confirming.

srowen commented on docs/configuration.md:
```diff
@@ -557,6 +557,20 @@ Apart from these, the following properties are also available, and may be useful
   </td>
 </tr>
+<tr>
+  <td><code>spark.worker.ui.retainedExecutors</code></td>
```
Although I don't like exposing yet more config knobs, these seem pretty harmless.

sarutak commented:
@JoshRosen Sure, I'll take a look. Have a good sleep :)

sarutak commented on core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:
```diff
@@ -614,6 +585,74 @@ private[worker] class Worker(
+
+  private def trimFinishedDriversIfNecessary(): Unit = {
+    if (finishedDrivers.size > retainedDrivers) {
+      finishedDrivers.take(math.max(finishedDrivers.size / 10, 1)).foreach {
```
I noticed `finishedExecutors` and `finishedDrivers` are never accessed by keys. They are used like `finishedDrivers.values`.
So, can we use `ListBuffer` for `finishedExecutors` and `finishedDrivers` and remove elements by `finishedDrivers.trimStart(...)`?

CodingCat replied:
we can do that, but at the cost of paying some performance cost when call `.remove(execId)`, is it worth doing that?

sarutak commented on core/src/test/scala/org/apache/spark/deploy/worker/WorkerSuite.scala:
```diff
@@ -17,13 +17,18 @@
+import org.apache.spark.deploy.master.DriverState
+import org.apache.spark.rpc.{RpcAddress, RpcEnv}
+import org.apache.spark.{SecurityManager, SparkConf, SparkFunSuite}
+import org.apache.spark.deploy.{ExecutorState, Command}
```
This should be `import org.apache.spark.deploy.{Command, ExecutorState}` .
Also, could you sort these import statements in dictionary order?

CodingCat replied:
sure, doing that

sarutak commented on core/src/test/scala/org/apache/spark/deploy/DeployTestUtils.scala:
```diff
@@ -0,0 +1,68 @@
+  }
+
+  def createExecutorRunner(execId: Int): ExecutorRunner = {
+    new ExecutorRunner("appId", execId, createAppDesc(), 4, 1234, null, "workerId", "host", 123,
```
Here and in any other places, when we put many parameters, we should write like as follows right?

```
new ExecutorRunner(
  "appId",
  execId,
  createAppDesc(),
  4,
  1234,
  null,
  "workerId",
  "host",
  123,
  ...)
```

CodingCat replied:
sure, addressing that

CodingCat committed:
addressing the comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38841 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38841/consoleFull) for   PR 7714 at commit [`54249ac`](https://github.com/apache/spark/commit/54249ac65e3caac84d0ca91771a9f9eb827635db).

SparkQA commented:
[Test build #38841 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38841/console) for   PR 7714 at commit [`54249ac`](https://github.com/apache/spark/commit/54249ac65e3caac84d0ca91771a9f9eb827635db).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

CodingCat committed:
add comments about why we don't synchronize finishedExecutors & finishedDrivers

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38933 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38933/consoleFull) for   PR 7714 at commit [`eb0f66e`](https://github.com/apache/spark/commit/eb0f66eacdbaf575d2bfd3bda0eb14db33013a38).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38944 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38944/consoleFull) for   PR 7714 at commit [`1b51a37`](https://github.com/apache/spark/commit/1b51a37fd4de8bbfddc8888d3d81da63ed561068).

SparkQA commented:
[Test build #38933 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38933/console) for   PR 7714 at commit [`eb0f66e`](https://github.com/apache/spark/commit/eb0f66eacdbaf575d2bfd3bda0eb14db33013a38).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

CodingCat commented:
some flaky tests....

SparkQA commented:
[Test build #38944 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38944/console) for   PR 7714 at commit [`1b51a37`](https://github.com/apache/spark/commit/1b51a37fd4de8bbfddc8888d3d81da63ed561068).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39041 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39041/consoleFull) for   PR 7714 at commit [`23977fb`](https://github.com/apache/spark/commit/23977fb3bc590f58e9d4d44cfcce78ce0a49baca).

SparkQA commented:
[Test build #39041 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39041/console) for   PR 7714 at commit [`23977fb`](https://github.com/apache/spark/commit/23977fb3bc590f58e9d4d44cfcce78ce0a49baca).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

CodingCat commented:
finally....@srowen, @JoshRosen, @sarutak more comments?

srowen commented:
It's looking good to me. Let's leave it open another day or two for comments.

srowen commented:
@JoshRosen @sarutak did you want to look again? otherwise I think this can go in today. I know there's a huge amount of traffic at the moment so wanted to check again

sarutak commented:
Yeah, I think this is ready to merge.

CodingCat commented:
@JoshRosen @sarutak @srowen  thanks!
